portal/models/questionnaire_bank.py

```diff
@@ -0,0 +1,176 @@
+        db.ForeignKey('organizations.id'), nullable=False)
+
+    def __str__(self):
+        """Print friendly format for logging, etc."""
+        return "QuestionnaireBank {0.id} {0.name}".format(self)
+
+    @classmethod
+    def from_json(cls, data):
+        instance = cls()
+        instance.update_from_json(data)
+        return instance
+
+    def update_from_json(self, data):
+        self.name = data['name']
+        self.organization_id = Reference.parse(
+            data['organization']).id
+        self.add_if_not_found(commit_immediately=True)
+        for q in data['questionnaires']:
+            questionnaire = QuestionnaireBankQuestionnaire.from_fhir(
+                q)
+            questionnaire.questionnaire_bank_id = self.id
+            questionnaire = questionnaire.add_if_not_found(True)
+
+    def as_json(self):
+        d = {}
+        d['resourceType'] = 'QuestionnaireBank'
+        d['name'] = self.name
+        d['organization'] = Reference.organization(
+            self.organization_id).as_fhir()
+        d['questionnaires'] = [q.as_fhir() for q in self.questionnaires]
+        return d
+
+    def add_if_not_found(self, commit_immediately=False):
+        """Add self to database, or return existing
+
+        Queries for similar, adds new if not found.
+
+        @return: the new or matched QuestionnaireBank
+
+        """
+        existing = QuestionnaireBank.query.filter_by(
+            name=self.name).first()
+        if not existing:
+            db.session.add(self)
+            if commit_immediately:
+                db.session.commit()
+        else:
+            self.id = existing.id
+        self = db.session.merge(self)
+        return self
+
+    @staticmethod
+    def q_for_user(user):
+        """Lookup and return all questionnaires for the given user
+
+        QuestionnaireBanks are associated with a user through the top
+        level organization affiliation.
+
+        :return: list of QuestionnaireBankQuestionnaire objects for given user
+
+        """
+        qs = []
+        OT = OrgTree()
+        for org in user.organizations:
+            # Only top level orgs named in associations w/ QuestionnairBanks
+            top = OT.find(org.id).top_level()
+            qb = QuestionnaireBank.query.filter_by(organization_id=top).first()
+            if qb:
+                qs += [q for q in qb.questionnaires]
+        return qs
+
+
+class QuestionnaireBankQuestionnaire(db.Model):
+    """link table for n:n association between Questionnaires and Banks"""
+    __tablename__ = 'questionnaire_bank_questionnaires'
+    id = db.Column(db.Integer(), primary_key=True)
+    questionnaire_bank_id = db.Column(db.Integer(), db.ForeignKey(
+        'questionnaire_banks.id', ondelete='CASCADE'), nullable=False)
+    questionnaire_id = db.Column(db.Integer(), db.ForeignKey(
+        'questionnaires.id', ondelete='CASCADE'), nullable=False)
+    days_till_due = db.Column(db.Integer, nullable=False)
```

Comment:
I don't believe so, at least not at this time.  Adding the constraint on the db means the comparison code doesn't have to handle nulls.  But perhaps I'm wrong...